### DISCUSSION TITLE
Using of stepper motors

### DISCUSSION THREAD

**Discussion by Kalli87** [AndreWohnsland/CocktailBerry#182] [Ideas]
Ich habe an meiner Cocktailmaschine Nema 17 Motoren verbaut.
Ist es möglich, das Programm so zu ändern, dass man diese Motoren damit betreiben kann? 
Leider habe ich noch nicht so viel Erfahrung mit dem Programmieren.

**Comment by AndreWohnsland** [comment]
Hi, die kurze Antwort ist: gerade nicht, und ist auch nicht ganz einfach zu machen, aber an sich geht das schon.

Die längere: Die aktuelle Implementation geht davon aus, dass jede Pumpe einen Pin am Pi belegt und einfach nur (durch Relais etc.) für eine gewisse Zeit angeschaltet wird, bis das Zutaten Volumen erreicht ist. Ich bin jetzt echt kein Experte bei Stepper Motoren, aber soweit ich das verstehe, haben die Controller ja min. 2 Pins, einen Richtungs-Pin und eine Stepper/Signal Pin. Zudem schaltet man diese ja nicht einfach an und am Ende aus, sondern gibt pro Step ein Signal.
Das ist dann schon eine andere Logik, sowie auch eine andere Anzahl an Konfigurationsparameter, die die Benutzeroberfläche auch nicht unterstützt.
Um jetzt nicht zu technisch zu werden: man könnte theoretisch den Support diverse „Pumpen“, also eine Relais, gesteuerte Pumpe, Stepper Motor gesteuerte Pumpe, etc. implementieren. Da sowas nicht vorgesehen war, geht das nicht ganz „so einfach“ beziehungsweise macht auch Probleme auf der grafischen Seite von den Konfigurationen.
Klar könntest du auch lokal den Code ändern, aber das braucht ein bisschen Code Verständnis.
Wie steuerst du die Motoren an, beziehungsweise was war der Grund für die Stepper Motoren? An sich sind die ja eigentlich mehr für sehr genaue Aktionen, was ja hier Overkill ist.

*EDIT*: Eine Sache ist, die ich mir noch vorstellen kann, dass man in den Optionen stattdessen ein Custom Script einbetten kann, welches eine Klasse hat, die ein gewisses Schema verfolgen muss und dann in das Programm geladen wird. Diese Klassen würden stattdessen die Pumpen steuern, anstelle des internen Codes in CocktailBerry. Das würde natürlich auch erfordern, dass man selbst (also die Maschinenbesitzer) für ihr abweichenden Pumpen die Implementation schreiben muss. Wäre aber auch ein Weg, ohne enorme Änderungen im "Base" Code so etwas zu ermöglichen. 

LG Andre

**Reply by Kalli87** [reply]
Erstmal vielen Dank für die ausführliche Antwort.

Die Steppermotoren habe ich genommen, da ich noch welche zuhause hatte. Ich konstruiere öfters technische Gegenstände, wie z.B. 3D Drucker und CNC-Fräse und jetzt auch die Cocktailmaschine mit selbstgebauten Peristaltikpumpen. Für die Maschine habe ich im Moment eine HTML basierte App in der ich auswählen kann ob ich Gleichstrommotoren oder Steppermotoren verwende. Dia App funktioniert wunderbar, hat jedoch das Problem, dass ich daran nichts ändern kann und somit keine Lichtsteuerung implementieren kann.
Das mit dem Overkill ist mir auch klar..., aber mein innerer Monk sieht das anders. ;-)

Ich muss mich da einfach mal rangeben und das Programmieren lernen. Bei meinen bisherigen Projekten konnte ich mir die Codes immer aus verschiedensten Projekten Zusammenkopieren, da die Cocktailmaschine eher eine seltenes Projekt ist, ist das hier nicht so leicht. Ich versuche es mal weiter, vielleicht finde ich ja noch den passenden Baustein zum implementieren.

Gruß Daniel

![tempImageYembkR](https://github.com/user-attachments/assets/f2b844f2-d906-4758-ac7e-fd271b63a617)
Die Maschine befindet sich noch im Aufbau, funktioniert aber schon einwandfrei.

**Reply by AndreWohnsland** [reply]
Hach ja, an selbst gebaute Peristaltikpumpen hatte ich da schon gedacht 😆.
Erinnert mich auch ein wenig an meine 3D Drucker ^^.
Hast du zufälligerweise den Python code zum ansteuern der Motoren? 
Hatte online ein paar gesehen, bin aber doch interessiert wie man das generell so machen kann :).

**Reply by Kalli87** [reply]
Finde da gerade kein passenden Code zu. Vor allem sind die Codes aus dem 3D Drucker zu Aufwändig. 
[RpiMotorLib](https://github.com/gavinlyonsrepo/RpiMotorLib/blob/master/RpiMotorLib/RpiMotorLib.py) Hier ist das ganz gut erklärt, auch mit Code
